core/src/utils.jl

```diff
@@ -55,6 +55,56 @@ function create_storage_tables(
     return area, level, storage
 end
 
+"""Compute the storages of the basins based on the water level of the basins."""
+function get_storages_from_levels(
+    basin::Basin,
+    levels::Vector{Float64},
+)::Tuple{Vector{Float64}, Bool}
+    storages = Float64[]
+    errors = false
+
+    for (i, level) in enumerate(levels)
+        storage_discrete = basin.storage[i]
+        area_discrete = basin.area[i]
+        level_discrete = basin.level[i]
+        bottom = first(level_discrete)
+
+        if level < bottom
+            node_id = basin.node_id[i]
+            @error "The level $level of basin #$node_id is lower than the bottom of this basin $bottom."
+            errors = true
+            push!(storages, NaN)
+            continue
+        end
+
+        level_lower_index = searchsortedlast(level_discrete, level)
+
+        # If the level is equal to the bottom then the storage is 0
+        if level_lower_index == 0
+            push!(storages, 0.0)
+            continue
+        elseif level_lower_index == length(level_discrete)
+            level_lower_index -= 1
+        end
+
+        darea =
+            (area_discrete[level_lower_index + 1] - area_discrete[level_lower_index]) /
```

Comment:
I think I'd prefer to see this packaged into a separate function, it's a nice logical unit. E.g. something like "interpolate_storage" or something:

```julia
function interpolate_storage(basin, level)
...
return storage
end

```

Reply:
I refactored it